crates/iota-types/src/stardust/address_swap_map.rs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashMap, str::FromStr};
+
+use serde::Deserialize;
+
+use crate::base_types::IotaAddress;
+type OriginAddress = IotaAddress;
+type DestinationAddress = IotaAddress;
+
+#[derive(Debug, Default, Deserialize)]
+pub struct AddressSwapMap {
+    addresses: HashMap<OriginAddress, DestinationAddress>,
+    address_swapped_at_least_once: HashMap<OriginAddress, bool>,
+}
+
+impl AddressSwapMap {
+    /// Retrieves the destination address for a given origin address.
+    /// Marks the origin address as swapped if found.
+    pub fn get_destination_address(
+        &mut self,
+        origin_address: OriginAddress,
+    ) -> Option<&DestinationAddress> {
+        self.addresses.get(&origin_address).inspect(|_| {
+            // Mark the origin address as swapped
+            if let Some(swapped_flag) = self.address_swapped_at_least_once.get_mut(&origin_address)
+            {
+                *swapped_flag = true;
+            }
+        })
+    }
```

Comment:
So, the fact that this takes a mut ref has some awkward implications for the verification. I would split this into one method to be used while migrating and one to simply fetch from the map. Also, the two hashmaps can be merged into one like

```rust
addresses: HashMap<OriginAddress, (DestinationAddress, Option<bool>)>,
```

or by creating a new struct like

```rust
struct DestinationAddress {
  address: IotaAddress,
  swapped: bool,
}

impl DestinationAddress {
  fn address(&self) -> IotaAddress {
    self.address
  }

  fn swap_address(&mut self) -> IotaAddress {
    self.swap = true;
    self.address()
  }
}
```

Reply:
I’ve refactored this part
Second hash map has been removed and also mut ref

Reply:
We need a mutable state to change this flag, but we don’t want to pass a mutable reference everywhere for the "getter". Is this what RefCell is meant for, or am I wrong?

Reply:
I mean, why are we changing this state during validation? I had assumed this was a flag that should change during migration

Reply:
`get_destination_address` calls during the migration.
We’re swapping addresses during the outputs migration.
For instance 

 ```
   pub(super) fn create_alias_objects(
        &mut self,
        header: &OutputHeader,
        alias: &AliasOutput,
        coin_type: CoinType,
        address_swap_map: &AddressSwapMap,
    ) -> Result<CreatedObjects> {
        let mut created_objects = CreatedObjects::default();

        // Take the Alias ID set in the output or, if its zeroized, compute it from the
        // Output ID.
        let alias_id = ObjectID::new(*alias.alias_id().or_from_output_id(&header.output_id()));
        let move_alias = iota_types::stardust::output::Alias::try_from_stardust(alias_id, alias)?;

        // TODO: We should ensure that no circular ownership exists.
        let alias_output_owner =
            **stardust_to_iota_address_owner_maybe_swap**(alias.governor_address(), address_swap_map)?;
            
 .......           
 }
```

Reply:
@DaughterOfMars I’ve merged your PR